pilot/pkg/proxy/envoy/v2/mesh_discovery.go

```diff
@@ -0,0 +1,219 @@
+	mesh MeshDiscovery
+	// grpcServer supports gRPC for xDS v2 services.
+	grpcServer *grpc.Server
+}
+
+func init() {
+	// TODO: move this to Pilot's configuration.
+	tempGRPCPort := os.Getenv("PILOT_GRPC_PORT")
+	portNum, err := strconv.Atoi(tempGRPCPort)
+	if err == nil && portNum > 1000 {
+		gRPCPort = tempGRPCPort
+	}
+}
+
+// IsEnabled returns true if EnvoyV2 APIs are enabled for Pilot.
+func IsEnabled() bool {
+	return len(gRPCPort) > 0
+}
+
+// NewDiscoveryServer creates DiscoveryServer that sources data from Pilot's internal mesh data structures
+func NewDiscoveryServer(mesh MeshDiscovery) *DiscoveryServer {
+	// TODO for now use hard coded / default gRPC options. The constructor may evolve to use interfaces that guide specific options later.
+	// Example:
+	//		grpcOptions = append(grpcOptions, grpc.MaxConcurrentStreams(uint32(someconfig.MaxConcurrentStreams)))
+	var grpcOptions []grpc.ServerOption
+
+	var interceptors []grpc.UnaryServerInterceptor
+
+	// TODO: log request interceptor if debug enabled.
+
+	// setup server prometheus monitoring (as final interceptor in chain)
+	interceptors = append(interceptors, grpc_prometheus.UnaryServerInterceptor)
+	grpc_prometheus.EnableHandlingTimeHistogram()
+
+	grpcOptions = append(grpcOptions, grpc.UnaryInterceptor(grpc_middleware.ChainUnaryServer(interceptors...)))
+
+	// get the grpc server wired up
+	grpc.EnableTracing = true
+
+	grpcServer := grpc.NewServer(grpcOptions...)
+	out := &DiscoveryServer{mesh: mesh, grpcServer: grpcServer}
+	xdsapi.RegisterEndpointDiscoveryServiceServer(out.grpcServer, out)
+	return out
+}
+
+// Start opens up a listening port for serving gRPC requests and starts up the underlying gRPC server.
+// This method is non-blocking.
+func (s *DiscoveryServer) Start() {
+	lis, err := net.Listen("tcp", ":"+gRPCPort)
+	if err != nil {
+		log.Errorf("mesh discovery server failed to listen on port %q: %v", gRPCPort, err)
+		return
+	}
+	// Serve is a blocking call.
+	go func(gRPCServer *grpc.Server, lis net.Listener) {
+		log.Infof("attempting to start mesh discovery server on port %q", gRPCPort)
+		if err := gRPCServer.Serve(lis); err != nil {
+			log.Errorf("mesh discovery server failed to serve on port %q: %v", gRPCPort, err)
+		}
+	}(s.grpcServer, lis)
+}
+
+/***************************  Mesh EDS Implementation **********************************/
+
+// StreamEndpoints implements xdsapi.EndpointDiscoveryServiceServer.StreamEndpoints().
+func (s *DiscoveryServer) StreamEndpoints(stream xdsapi.EndpointDiscoveryService_StreamEndpointsServer) error {
+	ticker := time.NewTicker(responseTickDuration)
+	peerInfo, ok := peer.FromContext(stream.Context())
+	peerAddr := "Unknown peer address"
+	if ok {
+		peerAddr = peerInfo.Addr.String()
+	}
+	defer ticker.Stop()
+	var discReq xdsapi.DiscoveryRequest
+	var receiveError error
+	var clusters []string
+	initialRequest := true
+	reqChannel := make(chan xdsapi.DiscoveryRequest, 1)
```

Comment:
This should probably be a channel of pointers to the DiscoveryRequest - protos are always pointers in golang, and we get a new copy per `Recv` so there's not a risk of a data race sharing the ptr.